flatten out testing logic to make tests more readable

vkarpov15 opened the pull request:
<!--- Note to EXTERNAL Contributors -->
<!-- Thanks for opening a PR! 
If it is a significant code change, please **make sure there is an open issue** for this. 
We work best with you when we have accepted the idea first before you code. -->

<!--- For ALL Contributors 👇 -->

## What was changed

Current code uses `runUntil()`, [which makes it a bit tricky to test queries and signals](https://github.com/temporalio/documentation/pull/1252). It could also benefit from using Jest hooks rather than helper functions.

Also, why are we using Jest here? I'm a bit wary of recommending users use Jest.

## Why?

I haven't been able to repro https://github.com/temporalio/sdk-typescript/issues/728, but I suspect that `runUntil()` may have something to do with it. @joebowbeer can you try out these tests and see if you're still experiencing the issue you reported in https://github.com/temporalio/sdk-typescript/issues/728 ?

## Checklist
<!--- add/delete as needed --->

1. Closes <!-- add issue number here -->

2. How was this tested:
<!--- Please describe how you tested your changes/how we can test them -->

3. Any docs updates needed?
<!--- update README if applicable
      or point out where to update docs.temporal.io -->

joebowbeer commented on position-node/src/workflows.test.ts:
```diff
@@ -1,10 +1,15 @@
+let runPromise: Promise<void>;
+let worker: Worker;
+
+const taskQueue = 'test';
```
```suggestion
const taskQueue = 'test';
const initInfo = gameInfo({ anchorX: 1 });
```

joebowbeer commented on position-node/src/workflows.test.ts:
```diff
@@ -18,9 +23,30 @@ beforeAll(async () => {
+
+beforeEach(async () => {
+  handle = await testEnv.workflowClient.start(pendulum, {
+    args: [gameInfo({ anchorX: 1 })],
```
```suggestion
    args: [initInfo],
```

joebowbeer commented on position-node/src/workflows.test.ts:
```diff
@@ -50,49 +76,20 @@ test('gameInfo with overrides', () => {
 });
 
 test('pendulum getGameInfoQuery', async () => {
   const info = gameInfo({ anchorX: 1 });
```
```suggestion
```

@vkarpov15 I like the refactor. It suits these tests well.

The one thing I don't like is that this info value is created out of thin air here. I think this should be hoisted to a constant `initInfo` where it can be used in `beforeEach` as a seed and then again in this test as an expected value.

joebowbeer commented on position-node/src/workflows.test.ts:
```diff
@@ -50,49 +76,20 @@ test('gameInfo with overrides', () => {
-    expect(queryResult).toEqual(info);
-  });
+  const queryResult = await handle.query(getGameInfoQuery);
+  expect(queryResult).toEqual(info);
```
```suggestion
  expect(queryResult).toEqual(initInfo);
```

joebowbeer reviewed (changes requested):
Thanks! The few changes suggested are hoisting the gameInfo _seed_ value.

Also: Why does each test leave its workflow dangling until the final afterAll? That seems unclean.

Previously, every test sent an exit signal and awaited a result. Perhaps that functionality can be moved into `afterEach`. (Then the exit-signal test can be an empty test.)

vkarpov15 commented:
@joebowbeer take a look at these latest changes, I just removed the `gameInfo()` function in favor of using the spread operator. I also added a `afterEach()` that terminates the Workflow handle to clean up Workflow handles.

I disagree with your assessment of Jest. In my experience, Jest should only be considered by expert architects with decades of JavaScript experience, because of the difficulty with configuring Jest and managing compatibility issues. The advantage of Mocha is that your code runs in plain old Node.js, whereas Jest replaces `require()`, `Date()`, and numerous other globals with their own implementations that are frequently incompatible with Node's. Jest isn't a testing framework, it's a whole separate JavaScript runtime. That's just my thoughts though, I don't speak for Temporal.

I'd be curious to hear from @bergundy and @lorensr about whether we should use Jest, Mocha, Ava, or something else as the "default" testing framework in our examples.

bergundy commented:
As users have pointed out, jest is by far the most popular testing framework for JS.
It's quite tricky to make the SDK work well with it and we've had at least 3-4 jest related issues reported already.
Ideally testing with jest would just work but in reality it looks like it's not that simple.

bergundy commented on position-node/src/workflows.test.ts:
```diff
@@ -18,81 +38,58 @@ beforeAll(async () => {
-      await handle.result();
+afterEach(async() => {
+  await handle.terminate().catch(err => {
+    // TODO: make TypeScript SDK export WorkflowNotFoundError
```
```
import { WorkflowNotFoundError } from '@temporalio/client';
```

bergundy reviewed (commented):
I personally prefer the first approach where we start a worker per test

vkarpov15 commented:
@bergundy why do you prefer starting a worker per test? Not saying you're wrong - I was just under the impression that starting a worker was an expensive operation, so it would be better to only start the worker once to make tests run faster.

@joebowbeer I agree that we should support Jest as much as possible, including having samples that use Jest. I'd just prefer that we use something lighter, like Mocha or Ava, in cases where we aren't specifically explaining testing with Jest or integrating with a framework whose community mostly uses Jest. For example, the Nest sample that I'm working on uses Jest, because that's what Nest's starter uses. Although for that, I'd like to add some documentation related to best practices about testing with Jest, similar to [Mongoose's Jest docs](https://mongoosejs.com/docs/jest.html), that mentions details like needing to set `NODE_OPTIONS=--experimental-vm-modules`

bergundy commented:
@vkarpov15 the reason I want to use a worker per test is because you typically mock activities per test and of course need a new worker for registration.

bergundy commented:
Starting a worker is pretty cheap, bundling workflows OTOH isn't so that is better done once per test file.

joebowbeer commented on position-node/src/workflows.test.ts:
```diff
@@ -18,81 +39,57 @@ beforeAll(async () => {
-      await handle.result();
+afterEach(async() => {
+  await handle.terminate().catch(err => {
+    if (err instanceof WorkflowNotFoundError) {
```
I assume this is only here to accommodate the exit signal test, but as a side-effect may also hide some unexpected  behavior.

Add comment?

joebowbeer commented on position-node/src/workflows.test.ts:
```diff
@@ -18,81 +39,57 @@ beforeAll(async () => {
-    const queryResult = await handle.query(getGameInfoQuery);
-    expect(queryResult).toEqual(info);
-  });
+  const info = { ...gameInfo, anchorX: 1 };
```
```suggestion
```

joebowbeer commented on position-node/src/workflows.test.ts:
```diff
@@ -18,81 +39,57 @@ beforeAll(async () => {
-  });
+  const info = { ...gameInfo, anchorX: 1 };
+  const queryResult = await handle.query(getGameInfoQuery);
+  expect(queryResult).toEqual(info);
```
```suggestion
  expect(queryResult).toEqual(initInfo);
```

The expected value of this query is not created out of thin air; it is the initial value provided on workflow creation.

joebowbeer commented on position-node/src/workflows.test.ts:
```diff
@@ -1,10 +1,31 @@
+  angleVelocity: 0,
+  dt: 0,
+  speed: 0,
+});
```
```suggestion
});

const initInfo = { ...gameInfo, anchorX: 1 };
```

joebowbeer commented on position-node/src/workflows.test.ts:
```diff
@@ -18,81 +39,57 @@ beforeAll(async () => {
-  expect(gameInfo({ anchorX: 1 }).anchorX).toBe(1);
+beforeEach(async () => {
+  handle = await testEnv.workflowClient.start(pendulum, {
+    args: [{ ...gameInfo, anchorX: 1 }],
```
```suggestion
    args: [initInfo],
```

joebowbeer reviewed (changes requested):
Looking good!

Request to hoist the initial value.

Also consider adding comment about ignored error.

vkarpov15 commented:
@joebowbeer I addressed your comments by checking the Workflow's status using `describe()` in `afterEach()`. What do you think of that approach?

@bergundy I made it so that the tests create a bundle using `bundleWorkflowCode()` in the `before()` hook, but create a Worker in `beforeEach()`. So create a new Worker for every test, but only bundle once. What do you think of that approach?

bergundy commented:
Creating the worker in `beforeEach` doesn't give any value over setting it up in the `before` hook.
My point for creating the worker in each test was so the worker can be customized if necessary (apparently it's not in this case), I'd be fine with either approach but in general in our samples I'd want to push users in the worker per test direction.

Anyways, TBH I don't see that much value in this PR, I think the code was fine the way it was but it's definitely not worse, just not a good use of time IMHO.

joebowbeerxealth reviewed (commented):
LGTM

joebowbeer reviewed (approved):
LGTM - but see most recent comment

vkarpov15 commented:
@joebowbeer I'm going to merge this PR since we have LGTMs. I think we're getting a little off track here, since this was just meant to be a quick PR to (1) potentially address https://github.com/temporalio/sdk-typescript/issues/728, (2) minimize callback-type code. I don't think us spending more time on this PR is valuable.
